PowerPointLabs/PowerPointLabs/SpeechEngine/TextToSpeech.cs

```diff
@@ -1,68 +1,174 @@
+            cortana.OnError += ErrorHandler;
+
+            // Reuse Synthesize object to minimize latency
+            cortana.Speak(CancellationToken.None, new Synthesize.InputOptions()
+            {
+                RequestUri = new Uri(requestUri),
+                Text = textToSpeak,
+                VoiceType = humanVoice.voiceType,
+                Locale = humanVoice.Locale,
+                VoiceName = humanVoice.voiceName,
+                // Service can return audio in different output format.
+                OutputFormat = AudioOutputFormat.Riff24Khz16BitMonoPcm,
+                AuthorizationToken = "Bearer " + accessToken,
+            }, filePath).Wait();
+        }
+
+        private static void SaveStringToWaveFile(String textToSave, String filePath)
+        {
+            PromptBuilder builder = GetPromptForText(textToSave);
+            PromptToAudio.SaveAsWav(builder, filePath);
+        }
+
+        private static string GetHumanSpeakNotesForText(string textToSave)
+        {
+            TaggedText taggedText = new TaggedText(textToSave);
+            string strToSpeak = taggedText.ToPrettyString();
+            return strToSpeak;
+        }
+        private static PromptBuilder GetPromptForText(string textToConvert)
+        {
+            TaggedText taggedText = new TaggedText(textToConvert);
+            PromptBuilder builder = taggedText.ToPromptBuilder(DefaultVoiceName);
+            return builder;
+        }
+
+        private static void SaveAudioToWaveFile(object sender, GenericEventArgs<Stream> args)
+        {
+            //  Logger.Log("saving " + args.FilePath);
+            SaveStreamToFile(args.FilePath, args.EventData);
+            args.EventData.Dispose();
+        }
+
+        private static byte[] ReadFully(Stream input)
+        {
+            byte[] buffer = new byte[16 * 1024];
+
+            using (MemoryStream ms = new MemoryStream())
+            {
+                int read;
+                while ((read = input.Read(buffer, 0, buffer.Length)) > 0)
+                {
+                    ms.Write(buffer, 0, read);
+                }
+                return ms.ToArray();
+            }
+        }
+        private static void SaveStreamToFile(string fileFullPath, Stream stream)
+        {
+            try
+            {
+                byte[] bytesInStream = ReadFully(stream);
+                using (FileStream fs = File.Create(fileFullPath))
+                {
+                    Console.WriteLine("file created");
+                    fs.Write(bytesInStream, 0, bytesInStream.Length);
+                }
+            }
+            catch (Exception e)
+            {
+                MessageBox.Show("Error generating audio files. ");
+                Debug.WriteLine(e.Message);
+            }
+        }
+        private static void ErrorHandler(object sender, GenericEventArgs<Exception> e)
+        {
+            Logger.Log("Unable to complete the TTS request: " + e.ToString());
```

Comment:
Just wondering, would it be better to display this to the user instead of just logger.log?

Reply:
I chose not to display this error to user because:
    1. In the case of http request failure due to max limit reached (response code 429), 3 retry attempts are handled by the code behind.
    2. In the case of network failure, the function will simply return and do nothing.

Both types of failures will fire `ErrorHandler` method. If user chooses to generate narrations for multiple slides, then multiple http requests will be sent out at one time because multiple threads are called to handle async requests. In the case of case 1 or case 2 error, multiple MessageBox will be displayed to user (because every request failure will fire `ErrorHandler` once). That would be undesirable.

Reply:
Oh I see.